README.md
# Smart-input-server

A simple server application to translate input from a smartphone onto a computer.  
The concept is simply to use smartphone as an input device for pointer (mouse) and keys (keyboard).

*The server part is done in Clojure because i want to practice with it for a while.*

Could be configured per application and shortcut buttons could be created.  
That is a basic idea for now...

## What's next

For now, Windows plateform is supported for the server.

The client (smart-phone app) will start with Android.  
It'll be another repository, I'll add a link when I have some commits to share.

## Usage

FIXME

## License

Copyright © 2017 Cedric Simon

GNU License v3

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-09-21
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-09-21
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/mouse_mover/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/mouse_mover/compare/0.1.0...0.1.1


test/mouse_mover/core_test.clj
(ns mouse-mover.core-test
  (:require [clojure.test :refer :all]
            [mouse-mover.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/mouse_mover/core.clj
(ns mouse-mover.core)

;; Ref: https://gist.github.com/stingh711/3760481
;; ==============================================

;;; UDP socket class
(import '(java.net DatagramSocket DatagramPacket))

;;; Rotot class
(import java.awt.Robot)
(import java.awt.MouseInfo)
(import java.awt.Toolkit)


;;; screen settings
(defn get-screen-size []
  "Return the screen dimensions as a typle {:width :height}"
  (let [dimensions (.getScreenSize (Toolkit/getDefaultToolkit))]
    {:width (.getWidth dimensions) :height (.getHeight dimensions)}
    ))


;;; ==================== mouse related ====================

(def robot (new Robot))

;;; current mouse position
(defn get-mouse-position []
  "Returns the mouse position on current screen as a tuple {:x :y}"
  (let [mouse-location (.getLocation (MouseInfo/getPointerInfo))]
    {:x (.x mouse-location) :y (.y mouse-location)}))


;;; moving the mouse on the screen
(defn set-mouse-position [x y]
  "Sets mouse cursor on screen"
  (.mouseMove robot x y))


;;; ==================== server related ===================

;; socket
(def socket (DatagramSocket. 5200))     ; need to learn the meaning of that writing style...
(def running (atom true))               ;can we change that @running at runtime?
(def buffer (make-array Byte/TYPE 1024))


;; handling packet data
(defn parse [packet]
  "TODO parse received messages and make action. First to come:
- moving the mouse"
  (let [message (String. (.getData packet) 0 (.getLength packet))]
    (println message)))



(defn start-server []
  ;; Do something about how to stop that infinite loop!
  (while (true? @running)
    (let [packet (DatagramPacket. buffer 1024)]
      (do
        (println "running")
        (.receive socket packet)
        (future (parse packet))))))


src/smart_input/server.clj
(ns smart-input.server)

;; Ref: https://gist.github.com/stingh711/3760481
;; ==============================================

;;; UDP socket class
(import '(java.net DatagramSocket DatagramPacket))

;;; Rotot class
(import java.awt.Robot)
(import java.awt.MouseInfo)
(import java.awt.event.InputEvent)
(import java.awt.Toolkit)
(import java.awt.event.KeyEvent)


;;; screen settings
(defn get-screen-size []
  "Return the screen dimensions as a typle {:width :height}"
  (let [dimensions (.getScreenSize (Toolkit/getDefaultToolkit))]
    {:width (.getWidth dimensions) :height (.getHeight dimensions)}
    ))


;;; ==================== mouse related ====================

(def robot (new Robot))

;;; current mouse position
(defn get-mouse-position []
  "Returns the mouse position on current screen as a tuple {:x :y}"
  (let [mouse-location (.getLocation (MouseInfo/getPointerInfo))]
    {:x (.x mouse-location) :y (.y mouse-location)}))


;;; moving the mouse on the screen
(defn set-mouse-position [x y]
  "Sets mouse cursor on screen"
  (.mouseMove robot x y))


;;; Keyboard related
;;; reading about Keys: https://docs.oracle.com/javase/7/docs/api/java/awt/event/KeyEvent.html
(defn key-press [data]
  "Simply simulates key presses.
TODO layouts needs to be handled differently for non alphanumeric characters!"
  (.keyPress   robot KeyEvent/VK_SHIFT)
  (.keyPress   robot KeyEvent/VK_4)       ;only one key for now: Shift+4 = $ (JIS layout)...
  (.keyRelease robot KeyEvent/VK_4)
  (.keyRelease robot KeyEvent/VK_SHIFT))


;;; ==================== server related ===================

;; socket
(def socket (DatagramSocket. 5200))     ; need to learn the meaning of that writing style...
(def running (atom true))               ;can we change that @running at runtime?
(def buffer (make-array Byte/TYPE 1024))


(defn parse-mouse-data-move [data]
  "Parsing mouse signal data and act accordingly.
Data format: X+n,Y+m where + can also be - and n and m are numbers.
What to do if n and m are not parseable? Actually an exception is raised. 
Return true if data are correctly formated for the operation.
False otherwise."
  (println "move mouse!" data)
  (let [[_ x-offset y-offset] (re-matches #"X([+-]\d+),Y([+-]\d+)" data)
        mouse-pos (get-mouse-position)]
    (set-mouse-position (+ (:x mouse-pos) (Integer. x-offset))
                        (+ (:y mouse-pos) (Integer. y-offset)))
    true))


(defn parse-mouse-data-click [data]
  "Makes the mouse click
TODO deal with different buttons, for now only left click (ie, data ignored)"
  (println "mouse click!" data)
  (.mousePress robot InputEvent/BUTTON1_MASK)
  (.mouseRelease robot InputEvent/BUTTON1_MASK)
  true)


;; handling packet data
(defn parse [packet]
  "WIP parse received messages and make action.
- MOUSE_MOVE :: moving the mouse
- MOUSE_CLICK :: click the mouse button (1:left)
- EXIT :: stoping the server from the client"
  (let [message (String. (.getData packet) 0 (.getLength packet))
        [_ signal data] (re-matches #"^([A-Z_]+):(.*)" message)]
    (println "received message >" message "<")

    ;; what messages do we handle?
    (case signal
      "MOUSE_MOVE" (parse-mouse-data-move data)
      "MOUSE_CLICK" (parse-mouse-data-click data)
      false) 
    ;; TODO add other signals
    ))


(defn -main []
  "Starting the udp server!"
  (reset! running true)
  (println "running")
  
  ;; waiting loop
  (while (true? @running)
    (let [packet (DatagramPacket. buffer 1024)]
      (do
        (.receive socket packet)
        (reset! running  @(future (parse packet)))
        ;; @ of a future will deference it and return the value of the computation!
        (println "new running state >" @running "<")
        )))
  (System/exit 0))
